File: app/src/main/java/code/name/monkey/retromusic/mvp/presenter/ArtistDetailsPresenter.java
package code.name.monkey.retromusic.mvp.presenter;

import android.os.Bundle;
import android.support.annotation.NonNull;

import code.name.monkey.retromusic.model.Artist;
import code.name.monkey.retromusic.mvp.Presenter;
import code.name.monkey.retromusic.mvp.contract.ArtistDetailContract;

import static code.name.monkey.retromusic.ui.activities.ArtistDetailActivity.EXTRA_ARTIST_ID;


/**
 * Created by hemanths on 20/08/17.
 */

public class ArtistDetailsPresenter extends Presenter implements ArtistDetailContract.Presenter {

    @NonNull
    private final ArtistDetailContract.ArtistsDetailsView view;
    private Bundle bundle;

    public ArtistDetailsPresenter(@NonNull ArtistDetailContract.ArtistsDetailsView view,
                                  Bundle artistId) {
        this.view = view;
        this.bundle = artistId;
    }

    @Override
    public void subscribe() {
        loadArtistById();
    }

    @Override
    public void unsubscribe() {
        disposable.clear();
    }

    @Override
    public void loadArtistById() {
        disposable.add(repository.getArtistById(bundle.getInt(EXTRA_ARTIST_ID))
                .subscribeOn(schedulerProvider.computation())
                .observeOn(schedulerProvider.ui())
                .doOnSubscribe(disposable1 -> view.loading())
                .subscribe(this::showArtist,
                        throwable -> view.showEmptyView(),
                        view::completed));
    }

    private void showArtist(Artist album) {
        if (album != null) {
            view.showData(album);
        } else {
            view.showEmptyView();
        }
    }
}

File: app/src/main/java/code/name/monkey/retromusic/mvp/presenter/ArtistDetailsPresenter.kt
/*
 * Copyright (c) 2019 Hemanth Savarala.
 *
 * Licensed under the GNU General Public License v3
 *
 * This is free software: you can redistribute it and/or modify it under
 * the terms of the GNU General Public License as published by
 *  the Free Software Foundation either version 3 of the License, or (at your option) any later version.
 *
 * This software is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU General Public License for more details.
 */

package code.name.monkey.retromusic.mvp.presenter

import code.name.monkey.retromusic.model.Artist
import code.name.monkey.retromusic.mvp.BaseView
import code.name.monkey.retromusic.mvp.Presenter
import code.name.monkey.retromusic.mvp.PresenterImpl
import code.name.monkey.retromusic.providers.interfaces.Repository
import code.name.monkey.retromusic.rest.model.LastFmArtist
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import java.util.*
import javax.inject.Inject
import kotlin.coroutines.CoroutineContext

/**
 * Created by hemanths on 20/08/17.
 */
interface ArtistDetailsView : BaseView {

    fun artist(artist: Artist)

    fun artistInfo(lastFmArtist: LastFmArtist?)

    fun complete()
}

interface ArtistDetailsPresenter : Presenter<ArtistDetailsView> {

    fun loadArtist(artistId: Int)

    fun loadBiography(
        name: String,
        lang: String? = Locale.getDefault().language,
        cache: String?
    )

    class ArtistDetailsPresenterImpl @Inject constructor(
        private val repository: Repository
    ) : PresenterImpl<ArtistDetailsView>(), ArtistDetailsPresenter, CoroutineScope {

        override val coroutineContext: CoroutineContext
            get() = Dispatchers.IO + job

        private val job = Job()

        override fun loadBiography(name: String, lang: String?, cache: String?) {
            launch {
                val result = repository.artistInfo(name, lang, cache)
                view?.artistInfo(result)
            }
        }

        override fun loadArtist(artistId: Int) {
            launch {
                val result = repository.artistById(artistId)
                view?.artist(result)
            }
        }

        override fun detachView() {
            super.detachView()
            job.cancel()
        }
    }
}